kong/plugins/prometheus/wasmx.lua

```diff
@@ -0,0 +1,210 @@
+local buffer = require "string.buffer"
+local wasm = require "kong.runloop.wasm"
+local wasmx_shm
+
+
+local fmt = string.format
+local str_find = string.find
+local str_sub = string.sub
+local table_insert = table.insert
+local table_sort = table.sort
+local buf_new = buffer.new
+local ngx_say = ngx.say
+
+
+local _M = {}
+
+
+local FLUSH_EVERY = 100
+
+
+local function sorted_iter(ctx, i)
+  i = i + 1
+
+  local v = ctx.t[ctx.sorted_keys[i]]
+
+  if v ~= nil then
+    return i, v
+  end
+end
+
+
+local function sorted_pairs(t)
+  local sorted_keys = {}
+
+  for k, _ in pairs(t) do
+    table_insert(sorted_keys, k)
+  end
+
+  table_sort(sorted_keys)
+
+  return sorted_iter, { t = t, sorted_keys = sorted_keys }, 0
+end
+
+--
+-- Convert a pw_key into a pair of metric name and labels
+--
+-- pw_key follows the form `pw.<filter_name>.<metric_name>`
+-- `<metric_name>` might contain labels, e.g. a_metric_label1="v1";
+-- if it does, the position of the first label corresponds to the end of the
+-- metric name and is used to discard labels from <metric_name>.
+local function parse_pw_key(pw_key)
+  local metric_name = pw_key
+  local _, _, filter_name = str_find(pw_key, "pw.(%a+)%.")
+  local filter_meta = wasm.filter_meta[filter_name] or {}
+  local label_patterns = {}
+  local labels = {}
+  local first_label_pos = #pw_key
+
+  if filter_meta.metrics and filter_meta.metrics.label_patterns then
+    label_patterns = filter_meta.metrics.label_patterns
+  end
+
+  for _, pair in ipairs(label_patterns) do
+    local label_pos, _, _, value = str_find(pw_key, pair.pattern)
```

Comment:
Given that this will be part of the public interface for filter developers, I would prefer to use `ngx.re.*` functions here instead of `string.find` so that our docs can point folks to PCRE instead of Lua patterns. What do you think?

Reply:
I think this is a good idea. Thanks for suggesting it.